HHH-10083 Support replicated and distributed caches

rvansa opened the pull request:
https://hibernate.atlassian.net/browse/HHH-10083

galderz commented on hibernate-infinispan/src/main/java/org/hibernate/cache/infinispan/access/TombstoneAccessDelegate.java:
```diff
@@ -0,0 +1,175 @@
+	public void remove(SessionImplementor session, Object key) throws CacheException {
+		TransactionCoordinator transactionCoordinator = session.getTransactionCoordinator();
+		TombstoneSynchronization sync = new TombstoneSynchronization(transactionCoordinator, asyncWriteCache, requiresTransaction, region, key);
+		transactionCoordinator.getLocalSynchronizations().registerSynchronization(sync);
```
The order in which the synchronization is registered is switched compared to `write`, is that on purpouse?

rvansa replied:
No, and actually the synchronization should be registered only after successful put(), because we don't want to run the synchronization if it fails -> we're probably going to rollback the transaction. Well spotted!

galderz commented on hibernate-infinispan/src/main/java/org/hibernate/cache/infinispan/access/TombstoneCallInterceptor.java:
```diff
@@ -0,0 +1,184 @@
+			}
+			else {
+				// two conflicting updates
+				setValue(e, new FutureUpdate(ZERO, null));
```
Are these cleared/expired eventually?

rvansa replied:
Hmm, if the current `FutureUpdate` is pre-, then yes, since the metadata is always set from the last put. But I forgot to add the expiration to the post-, good point. Btw., if you'd like to propose a better name, go on.

galderz commented:
@rvansa Great work, I'm wondering if the pending puts logics and this tombstone work could be merged into a single solution since their aim is similar: making sure older data does not override newer one?

I'm also wondering whether there's some tests that verify that no tombstones, tombstone updates or future updates are leaked, particularly in caches where no max size eviction is set and instead relies on per entity max lifespan/maxIdle time. 

Also, the other thing to consider is whether entries with tombstones, tombstone updates or future updates are considered as being counting entries when it comes to cache sizes and the size limits upon which eviction is calculated. For example, tombstones should not count towards the max size cache since they're just temporarily stored for a duration of time.

rvansa commented:
@galderz If we want to merge those two - how? The major difference is that PP use another cache for its metadata, while tombstones store all in one cache. Do we want to move PP info into entity cache? I think that even if we've done that, there would be non-trivial differences as PP needs to send more messages than regular invalidation mode. Though, I agree that the logic is very similar.

Good point about hunting down the leaked resources, I will add some tests for that. I should also either make sure that eviction does not remove the tombstones (if that's possible), or add a check that it is not enabled for tombstone caches.

rvansa commented:
@galderz Hi Galder, I've added those tests and fixed the flaws. Please squash the commit, I've separated that for easier review.

When trying out the tests, I've found out that some of the situations handled in the code cannot happen as DB prohibits some conflicts itself, so the test coverage may not be 100%.

rvansa commented:
@galderz One more thing I should check with you: the updates allow stale value read before the transaction finished, something I've mentioned in my last mail to hibernate-dev and described a bit in https://github.com/infinispan/infinispan/wiki/Infinispan-Hibernate-Second-Level-Cache-improvements#consistency-fixes , too. I am considering if I should make the behaviour strict (removing the value when the update is about to happen) and expose the stale values only in the non-strict strategy.

rvansa commented on hibernate-infinispan/src/main/java/org/hibernate/cache/infinispan/access/TombstoneCallInterceptor.java:
```diff
@@ -0,0 +1,196 @@
+		Map<Object, CacheEntry> contextEntries = ctx.getLookedUpEntries();
+		AdvancedCache decoratedCache = cache.getAdvancedCache().withFlags(flags != null ? flags.toArray(new Flag[flags.size()]) : null);
+		// In non-transactional caches we don't care about context
+		CloseableIterable<CacheEntry<Object, Void>> iterable = decoratedCache
```
Err, the generics are wrong here...

galderz commented:
> @galderz If we want to merge those two - how? The major difference is that PP use another cache for its metadata, while tombstones store all in one cache. Do we want to move PP info into entity cache? I think that even if we've done that, there would be non-trivial differences as PP needs to send more messages than regular invalidation mode. Though, I agree that the logic is very similar.

Ok, I haven't looked at the gory details of this but it was just a thought, in case at some point it can be consolidated in a single piece of logic.

> ... I should also either make sure that eviction does not remove the tombstones (if that's possible), or add a check that it is not enabled for tombstone caches.

That will work as long as you can guarantee that tombstones will eventually be removed without the need to have eviction turned on.

> @galderz One more thing I should check with you: the updates allow stale value read before the transaction finished, something I've mentioned in my last mail to hibernate-dev and described a bit in https://github.com/infinispan/infinispan/wiki/Infinispan-Hibernate-Second-Level-Cache-improvements#consistency-fixes , too. I am considering if I should make the behaviour strict (removing the value when the update is about to happen) and expose the stale values only in the non-strict strategy.

^ That makes sense.

rvansa commented:
@galderz Steve seemed that the staleness I've allowed here is not desired in the end, so I'll have that only in the non-strict mode. I'll mark this as preview then, and add commit with the fix soon (now I'm dived into the nonstrict mode).

rvansa commented:
Hi @galderz, I've issued the last fixes.

galderz commented:
Integrated, thanks @rvansa :)

sebersole commented:
Jira not closed.  Work not ported to 5.0 branch.

On Wed, Sep 16, 2015 at 7:55 AM Galder Zamarreño <notifications@github.com>
wrote:

> Closed #1074 <https://github.com/hibernate/hibernate-orm/pull/1074>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/hibernate/hibernate-orm/pull/1074#event-410921591>.
>

galderz commented:
Done
